### PR TITLE
feat: observer rewards

### PR DESCRIPTION
**Pull request by kingpinXD**
# Description
- Add functions to `testutil/network` for adding `crosschain` and `observer` genesis state, which tests can use in all modules. Refactor existing tests to use these functions instead of local counterparts in the module itself
- Add begin block function in `emissions` module to calculate and allocate rewards 
- Move `Block Reward components` to its file under `emissons/keeper` 
- Add ballot creation height to `ballots` .
- Add `BallotMaturityHeight` param .
- Add unit tests for `BuildRewardsDistribution` and  `DistributeObserverRewards`

- This PR does not add a migration script to modify existing ballots on the testnet , or distribute any of the older rewards 

Closes: #839 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

### CODE REVIEW

File: x/emissions/abci.go [hunk 1]
```diff
@@ -35,71 +37,92 @@ func BeginBlocker(ctx sdk.Context, keeper keeper.Keeper, stakingKeeper types.Sta
...
+	for _, ballotIdentifier := range ballotIdentifiers {
+		ballot, found := keeper.GetObserverKeeper().GetBallot(ctx, ballotIdentifier)
+		if !found {
+			continue
+		}
+		totalRewardsUnits = totalRewardsUnits + ballot.BuildRewardsDistribution(rewardsDistributer)
+	}
+	rewardPerUnit := sdkmath.ZeroInt()
+	if totalRewardsUnits > 0 && amount.IsPositive() {
+		rewardPerUnit = amount.Quo(sdk.NewInt(totalRewardsUnits))
```
Comment (kingpinXD) [inline x/emissions/abci.go:72]: Ballots can be cleared after this distribution if needed, decided not to do this in this PR . Can be added in later

**Review by lumtis — CHANGES_REQUESTED** [review]
👍

Many comments but many of them are small.

File: proto/emissions/withdrawble_emisions.proto [hunk 2]
```diff
@@ -0,0 +1,14 @@
+syntax = "proto3";
```
Comment (kevinssgh) [inline proto/emissions/withdrawble_emisions.proto:1]: Typo in file name: withdrawable_emissions.proto

File: x/crosschain/keeper/keeper_tss_voter.go [hunk 3]
```diff
@@ -53,14 +54,16 @@ func (k msgServer) CreateTSSVoter(goCtx context.Context, msg *types.MsgCreateTSS
...
 		}
 		ballot = observerTypes.Ballot{
-			Index:            "",
-			BallotIdentifier: index,
-			VoterList:        voterList,
-			Votes:            observerTypes.CreateVotes(len(voterList)),
-			ObservationType:  observerTypes.ObservationType_TSSKeyGen,
-			BallotThreshold:  sdk.MustNewDecFromStr("1.00"),
-			BallotStatus:     observerTypes.BallotStatus_BallotInProgress,
+			Index:                "",
```
[the remark this answers is not in the document]
Comment (kevinssgh) [reply x/crosschain/keeper/keeper_tss_voter.go:57]: The index is used when setting the ballot:

func (k Keeper) SetBallot(ctx sdk.Context, ballot *types.Ballot) {
	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.VoterKey))
	ballot.Index = ballot.BallotIdentifier

File: x/observer/types/ballot_test.go [hunk 4]
```diff
@@ -102,3 +102,56 @@ func TestVoter_IsBallotFinalized(t *testing.T) {
...
+				"Observer3": 1,
+				"Observer4": -1,
+			},
+		},
+		{
+			name:         "BallotFinalized_FailureObservation",
+			voterList:    []string{"Observer1", "Observer2", "Observer3", "Observer4"},
+			votes:        []VoteType{VoteType_SuccessObservation, VoteType_SuccessObservation, VoteType_FailureObservation, VoteType_FailureObservation},
+			ballotStatus: BallotStatus_BallotFinalized_FailureObservation,
+			expectedMap: map[string]int64{
```
Comment (kevinssgh) [inline x/observer/types/ballot_test.go:131]: Maybe I'm mistaken but the votes don't seem to match the expected map. Was that intentional?

File: proto/observer/params.proto [hunk 5]
```diff
@@ -56,4 +56,5 @@ message Params {
   option (gogoproto.goproto_stringer) = false;
   repeated ObserverParams observer_params =1;
   repeated Admin_Policy admin_policy= 2;
+  int64 ballot_maturity_blocks = 3;
```
Comment (kevinssgh) [inline proto/observer/params.proto:59]: I was curious what this maturity was for? Does the ballot have to be finalized for a certain number of blocks?

**Review by kevinssgh — COMMENTED** [review]
Looks good, just a few comments.

File: x/emissions/abci.go [hunk 6]
```diff
@@ -1,30 +1,32 @@
...
 	if err != nil {
 		panic(err)
 	}
-	err = DistributeObserverRewards(ctx, observerRewards, bankKeeper, observerKeeper)
+	err = DistributeObserverRewards(ctx, observerRewards, keeper)
 	if err != nil {
 		panic(err)
 	}
-	err = DistributeTssRewards(ctx, tssSignerRewards, bankKeeper)
+	err = DistributeTssRewards(ctx, tssSignerRewards, keeper.GetBankKeeper())
```
[the remark this answers is not in the document]
Comment (kingpinXD) [reply x/emissions/abci.go:29]: That is correct; a panic here would cause the chain to halt.
As far as I remember, the panic was added as a placeholder until we decide what to do here. 
Added a issue to remove it before mainnet : https://github.com/zeta-chain/node/issues/941

File: x/emissions/abci.go [hunk 7]
```diff
@@ -35,71 +37,92 @@ func BeginBlocker(ctx sdk.Context, keeper keeper.Keeper, stakingKeeper types.Sta
...
 
-func DistributeTssRewards(ctx sdk.Context, amount sdk.Int, bankKeeper types.BankKeeper) error {
-	coin := sdk.NewCoins(sdk.NewCoin(config.BaseDenom, amount))
-	return bankKeeper.SendCoinsFromModuleToModule(ctx, types.ModuleName, types.UndistributedTssRewardsPool, coin)
-}
-func GetBondFactor(ctx sdk.Context, stakingKeeper types.StakingKeeper, keeper keeper.Keeper) sdk.Dec {
-	targetBondRatio := sdk.MustNewDecFromStr(keeper.GetParams(ctx).TargetBondRatio)
-	maxBondFactor := sdk.MustNewDecFromStr(keeper.GetParams(ctx).MaxBondFactor)
-	minBondFactor := sdk.MustNewDecFromStr(keeper.GetParams(ctx).MinBondFactor)
+func DistributeObserverRewards(ctx sdk.Context, amount sdkmath.Int, keeper keeper.Keeper) error {
```
[the remark this answers is not in the document]
Comment (kingpinXD) [reply x/emissions/abci.go:50]: I agree; this should have some unit tests. 
However, it does get tested by simulating blocks here : https://github.com/zeta-chain/zeta-node/blob/6f7159e349e856e6123f9b64b4491320b2b0821b/x/emissions/client/tests/observer_rewards_test.go#L17-L17

File: x/emissions/abci.go [hunk 8]
```diff
@@ -35,71 +37,92 @@ func BeginBlocker(ctx sdk.Context, keeper keeper.Keeper, stakingKeeper types.Sta
...
+
+		if observerRewardsRatio == 0 {
+			finalDistributionList = append(finalDistributionList, &types.ObserverEmission{
+				EmissionType:    types.EmissionType_Slash,
+				ObserverAddress: observerAddress.String(),
+				Amount:          sdkmath.ZeroInt().String(),
+			})
+			continue
+		}
+		if observerRewardsRatio < 0 {
```
[the remark this answers is not in the document]
Comment (kingpinXD) [reply x/emissions/abci.go:97]: Modified the identifier to a more meaningful name , and added come comments to explain

File: x/emissions/abci.go [hunk 9]
```diff
@@ -35,71 +37,92 @@ func BeginBlocker(ctx sdk.Context, keeper keeper.Keeper, stakingKeeper types.Sta
...
+				finalDistributionList = append(finalDistributionList, &types.ObserverEmission{
+					EmissionType:    types.EmissionType_Slash,
+					ObserverAddress: observerAddress.String(),
+					Amount:          slashAmount.String(),
+				})
+			}
+			continue
+		}
+		// Defensive check
+		if rewardPerUnit.GT(sdk.ZeroInt()) {
```
[the remark this answers is not in the document]
Comment (kingpinXD) [reply x/emissions/abci.go:110]: All the conditions are not the same. Variable, I feel using a switch here might make the conditions more complicated than they need to be

File: x/emissions/abci.go [hunk 10]
```diff
@@ -35,71 +37,92 @@ func BeginBlocker(ctx sdk.Context, keeper keeper.Keeper, stakingKeeper types.Sta
...
+		if err != nil {
+			continue
+		}
+		observerRewardsRatio := rewardsDistributer[key]
+
+		if observerRewardsRatio == 0 {
+			finalDistributionList = append(finalDistributionList, &types.ObserverEmission{
+				EmissionType:    types.EmissionType_Slash,
+				ObserverAddress: observerAddress.String(),
+				Amount:          sdkmath.ZeroInt().String(),
```
[the remark this answers is not in the document]
Comment (kingpinXD) [reply x/emissions/abci.go:93]: Not really, do you mean just use "0"? 
Imo `sdkmath.ZeroInt().String()`, looks a bit more readable, considering the other value are all in type sdkmath.Int

File: x/emissions/keeper/block_rewards_components.go [hunk 11]
```diff
@@ -0,0 +1,61 @@
+package keeper
```
[the remark this answers is not in the document]
Comment (kingpinXD) [reply x/emissions/keeper/block_rewards_components.go:1]: Although we don't have individual tests , we do have a function to test them together 
[func TestAppModule_GetBlockRewardComponents(t *testing.T) ](https://github.com/zeta-chain/zeta-node/blob/6f7159e349e856e6123f9b64b4491320b2b0821b/x/emissions/abci_test.go#L76-L76)

[the remark this answers is not in the document]
Comment (kingpinXD) [reply x/emissions/keeper/block_rewards_components.go:1]: https://github.com/zeta-chain/node/issues/943

File: x/emissions/keeper/block_rewards_components.go [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
```
[the remark this answers is not in the document]
Comment (kingpinXD) [reply x/emissions/keeper/block_rewards_components.go:4]: Thanks for pointing this out 
sdk.Dec is replaced by sdk.LegacyDec in `cosmossdk.io/math.`  
This PR moves the code to a new file, we can move to using `cosmossdk.io/math.`, in a separate PR. Added an issue for it https://github.com/zeta-chain/node/issues/944

File: x/observer/types/ballot.go [hunk 13]
```diff
@@ -67,3 +70,32 @@ func CreateVotes(len int) []VoteType {
 	}
 	return voterList
 }
+
+// BuildRewardsDistribution builds the rewards distribution map for the ballot
+// It returns the total rewards units which account for the observer block rewards
+func (m Ballot) BuildRewardsDistribution(rewardsMap map[string]int64) int64 {
```
[the remark this answers is not in the document]
Comment (kingpinXD) [reply x/observer/types/ballot.go:76]: Te map is not saved to the storage 
Also , just to be safe it also sorted before use :https://github.com/zeta-chain/zeta-node/blob/6f7159e349e856e6123f9b64b4491320b2b0821b/x/emissions/abci.go#L79-L82

File: x/observer/keeper/ballot.go [hunk 14]
```diff
@@ -38,6 +54,50 @@ func (k Keeper) GetAllBallots(ctx sdk.Context) (voters []*types.Ballot) {
...
+	for ; iterator.Valid(); iterator.Next() {
+		var val types.Ballot
+		k.cdc.MustUnmarshal(iterator.Value(), &val)
+		if val.BallotStatus != types.BallotStatus_BallotInProgress {
+			store.Delete([]byte(val.Index))
+		}
+	}
+}
+
+func (k Keeper) AddBallotToList(ctx sdk.Context, ballot types.Ballot) {
```
[the remark this answers is not in the document]
Comment (kingpinXD) [reply x/observer/keeper/ballot.go:84]: Removing these functions as they are not used. It was originally planned to distribute rewards based on finalized ballots, but we moved away to matured ballots to account for the votes which are added after the ballot is finalized

File: x/observer/types/ballot_test.go (hunk 4 above)
Comment (kingpinXD) [reply x/observer/types/ballot_test.go:131]: Not sure I understand, the observer vote needs to match the finalized ballot status to get a rewards 
https://github.com/zeta-chain/zeta-node/blob/6f7159e349e856e6123f9b64b4491320b2b0821b/x/observer/types/ballot.go#L76-L76

File: x/observer/genesis.go [hunk 15]
```diff
@@ -21,7 +21,11 @@ func InitGenesis(ctx sdk.Context, k keeper.Keeper, genState types.GenesisState)
 	for _, elem := range genState.NodeAccountList {
 		k.SetNodeAccount(ctx, *elem)
 	}
-	k.SetParams(ctx, types.DefaultParams())
+	params := types.DefaultParams()
+	if genState.Params != nil {
```
[the remark this answers is not in the document]
Comment (kingpinXD) [reply x/observer/genesis.go:25]: It's not mandatory, do you see any reason we should not ?

File: proto/emissions/events.proto [hunk 16]
```diff
@@ -1,11 +1,25 @@
...
 
+enum EmissionType {
+    option (gogoproto.goproto_enum_stringer) = true;
+    Slash = 0;
+    Rewards = 1;
+}
+message ObserverEmission {
+    EmissionType emission_type = 1;
+    string observer_address = 2;
+    string amount = 3;
```
[the remark this answers is not in the document]
Comment (kingpinXD) [reply proto/emissions/events.proto:16]: Not sure I understand the question
The event should display the string representation of the big.Int?

File: proto/observer/ballot.proto [hunk 17]
```diff
@@ -30,5 +30,11 @@ message Ballot {
     (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Dec",
     (gogoproto.nullable) = false
   ];
-  BallotStatus ballot_status = 7;
+  BallotStatus ballot_status =7;
+  int64 Ballot_creation_height=8;
```
Comment (brewmaster012) [inline proto/observer/ballot.proto:34]: space, also naming convention should be at least not mixed upper case and underscore

File: proto/observer/genesis.proto [hunk 18]
```diff
@@ -18,4 +18,5 @@ message GenesisState {
   PermissionFlags permissionFlags = 4;
   Params params = 5;
   Keygen keygen = 6;
+  LastObserverCount lastObserverCount = 7;
```
Comment (brewmaster012) [inline proto/observer/genesis.proto:21]: i would do last_observer_count, permission_flags, 
let's use lowercase underscore for protobuf fields, 
let's try to follow protobuf style guide:
https://protobuf.dev/programming-guides/style/

File: proto/observer/params.proto [hunk 19]
```diff
@@ -48,12 +48,14 @@ enum Policy_Type {
 
 message Admin_Policy {
   Policy_Type policy_type = 1;
-  string address = 2;
+  string address =2 ;
```
Comment (brewmaster012) [inline proto/observer/params.proto:51]: space

**Review by brewmaster012 — COMMENTED** [review]
Overall looks good. 
Q: where is the performance of observer votes tracked and rewarded/slashed?

File: proto/emissions/events.proto (hunk 16 above)
[the remark this answers is not in the document]
Comment (lumtis) [reply proto/emissions/events.proto:16]: Basically why don't we directly set the type of `amount` as `Int`? It seems we do the conversion in the code

File: proto/emissions/query.proto [hunk 20]
```diff
@@ -20,6 +20,16 @@ service Query {
...
 
+// Queries a list of GetEmmisonsFactors items.
+	rpc GetEmmisonsFactors(QueryGetEmmisonsFactorsRequest) returns (QueryGetEmmisonsFactorsResponse) {
+		option (google.api.http).get = "/zeta-chain/zetacore/emissions/get_emmisons_factors";
+	}
+
+// Queries a list of ShowAvailableEmissions items.
+	rpc ShowAvailableEmissions(QueryShowAvailableEmissionsRequest) returns (QueryShowAvailableEmissionsResponse) {
+		option (google.api.http).get = "/zeta-chain/zetacore/emissions/show_available_emissions/{address}";
+	}
```
[the remark this answers is not in the document]
Comment (lumtis) [reply proto/emissions/query.proto:30]: Any thought regarding this comment?

File: x/emissions/abci.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (lumtis) [reply x/emissions/abci.go:100]: I meant if the type is `Int` we could you `sdkmath.ZeroInt()`

File: x/emissions/abci.go [hunk 21]
```diff
@@ -35,71 +37,92 @@ func BeginBlocker(ctx sdk.Context, keeper keeper.Keeper, stakingKeeper types.Sta
...
 	}
-	if fractionNumerator.IsZero() {
-		return sdk.ZeroDec()
+	sort.Strings(sortedKeys)
+
+	var finalDistributionList []*types.ObserverEmission
+	for _, key := range sortedKeys {
+		observerAddress, err := sdk.AccAddressFromBech32(key)
+		if err != nil {
+			continue
```
[the remark this answers is not in the document]
Comment (lumtis) [reply x/emissions/abci.go:90]: Thought?

File: x/emissions/keeper/block_rewards_components.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (lumtis) [reply x/emissions/keeper/block_rewards_components.go:1]: Yeah this is great for sure but unit tests allow to test edge cases for the specific method

File: x/observer/genesis.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (lumtis) [reply x/observer/genesis.go:25]: Might add a little more complexity but it is fine

File: x/observer/types/ballot.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (lumtis) [reply x/observer/types/ballot.go:77]: > Te map is not saved to the storage
You can still have issue if the iteration of the map influence the state

File: proto/emissions/query.proto (hunk 20 above)
[the remark this answers is not in the document]
Comment (kingpinXD) [reply proto/emissions/query.proto]: I couldn't find the command in our makefile.
Added one now to format the proto files on `make proto`

File: proto/observer/ballot.proto (hunk 17 above)
Comment (kingpinXD) [reply proto/observer/ballot.proto]: Added a proto formatter to make file and fixed the names

**Comment by kingpinXD** [comment]
> Overall looks good. Q: where is the performance of observer votes tracked and rewarded/slashed?

We are not tracking the performance. Is there a reason to do that? The ballots are not deleted, so we could write an off-chain solution to create that if needed.

Lets discuss this over a GitHub issue if needed

**Comment by brewmaster012** [comment]
> 

Ok we can do that separately.

File: x/emissions/abci.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (kingpinXD) [reply x/emissions/abci.go:90]: I already added a log line here 
https://github.com/zeta-chain/zeta-node/blob/0d2dce7a526870847899c267ca65989d06d40618/x/emissions/abci.go#L89-L89

Don't think we need to anything other than the log line though , since the chances of this error happening are impossible , as we do not have any way to add observers to the network , other than an upgrade

File: proto/emissions/query.proto (hunk 20 above)
[the remark this answers is not in the document]
Comment (lumtis) [reply proto/emissions/query.proto]: Mmh it got removed apparently https://github.com/zeta-chain/node/commit/e3f715e857aebaedb0d540741cb66362decb9bdf#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52L146

Any reason for this @fadeev ?

File: x/emissions/abci.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (lumtis) [reply x/emissions/abci.go:90]: Oh ok sorry, wasn't synchronized on my window 

> since the chances of this error happening are impossible

I think it's a valid reason to have a log because it would mean something is very wrong

**Review by lumtis — APPROVED** [review]
👍

**Review by brewmaster012 — APPROVED** [review]

File: proto/observer/genesis.proto [hunk 22]
```diff
@@ -15,7 +15,8 @@ message GenesisState {
   repeated Ballot ballots = 1;
   repeated ObserverMapper observers = 2;
   repeated NodeAccount nodeAccountList = 3;
```
Comment (ws4charlie) [inline proto/observer/genesis.proto:17]: put snake case on this field as well?

**Review by ws4charlie — APPROVED** [review]
